yt_dlp/downloader/external.py

```diff
@@ -302,6 +316,137 @@ def _make_cmd(self, tmpfilename, info_dict):
+             Popen(cmd, stdout=so.fileno(), stderr=se.fileno()) as p:
+            # make a small wait so that RPC client can receive response,
+            # or the connection stalls infinitely
+            time.sleep(0.2)
+            retval = p.poll()
+            while retval is None:
+                try:
+                    # https://aria2.github.io/manual/en/html/libaria2.html#c.DOWNLOAD_WAITING
+                    # https://aria2.github.io/manual/en/html/aria2c.html#aria2.tellActive
+
+                    # use tellActive as we won't know the GID without reading stdout
+                    # that is a mess in Python
+                    aktiva = aria2c_rpc('aria2.tellActive', [])
+                    completed = aria2c_rpc('aria2.tellStopped', [0, abs(nr_frags)])
+                    if not aktiva and len(completed) == abs(nr_frags):
+                        # no active downloads, we'll exit the loop after shutdown
+                        aria2c_rpc('aria2.shutdown', [])
+                        retval = p.wait()
+                        break
+
+                    if nr_frags < 0:
+                        # single file
+                        active = aktiva[0]
+                        cl, ds, tl = int(active['completedLength']), int(active['downloadSpeed']), int(active['totalLength'])
+                        status.update({
+                            'downloaded_bytes': cl,
+                            'speed': ds,
+                            'eta': try_get(0, lambda x: (tl - cl) / ds),
+                            'total_bytes': tl,
+                        })
+                        continue
+
+                    # fragmented
+                    total_bytes = sum(map(int, traverse_obj([aktiva, completed], (..., ..., 'totalLength'), default=[])))
+                    if completed or aktiva:
+                        total_bytes = total_bytes * nr_frags / (len(completed) + len(aktiva))
+                    total_completed = sum(map(int, traverse_obj(completed, (..., 'totalLength'), default=[])))
+                    dled_aktiva = sum(map(int, traverse_obj(aktiva, (..., 'completedLength'), default=[])))
+                    total_speed = sum(map(float, traverse_obj(aktiva, (..., 'downloadSpeed'), default=[])))
+                    dl_all = dled_aktiva + total_completed
+
+                    status.update({
+                        'downloaded_bytes': dl_all,
+                        'speed': total_speed,
+                        'eta': try_get(0, lambda x: (total_bytes - dl_all) / total_speed),
+                        'total_bytes': total_bytes,
+                        'fragment_index': len(completed) + len(aktiva) // 2,
+                    })
+                finally:
+                    status.update({'elapsed': time.time() - started})
+                    self._hook_progress(status, info_dict)
+                    time.sleep(0.1)
+                    retval = p.poll()
```

Comment:
why is this in `finally`? especially sleep and poll

Reply:
The main places where these lines are needed; at the end of `if nr_frags < 0:` and right before this `finally`
I guess it's better to have it in `finally` as it always run even after `continue`